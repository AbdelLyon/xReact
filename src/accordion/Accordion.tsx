import { ReactNode } from "react";

import {
  Accordion as AccordionRoot,
  AccordionContent,
  AccordionItem,
  AccordionTrigger,
} from "@/ui/accordion";
import { cn } from "@/utils";

export type AccordionItemData = {
  title: () => ReactNode;
  description?: () => ReactNode;
  value: string;
  content: () => ReactNode;
  className?: string;
};

type AccordionProps = {
  classNames?: {
    container?: string;
    item?: string;
    title?: string;
    content?: string;
    description?: string;
  };
  items: AccordionItemData[];
  renderContent?: (item: AccordionItemData) => ReactNode;
  renderHeader?: (item: AccordionItemData) => ReactNode;
};

export const Accordion: React.FC<AccordionProps> = ({
  classNames,
  items,
  renderContent,
  renderHeader,
}) => {
  return (
    <div className={cn(classNames?.container)}>
      {items.map(({ title, description, value, content, className }) => (
        <AccordionRoot
          key={value}
          collapsible
          className={cn(
            "w-full rounded-md bg-card-light dark:bg-card",
            classNames?.item,
            className,
          )}
          type="single"
        >
          <AccordionItem
            className={cn("border border-border rounded-md")}
            value={value}
          >
            <AccordionTrigger className={cn("pr-3 pl-1", classNames?.title)}>
              {renderHeader ? (
                renderHeader({ title, description, value, content })
              ) : (
                <>
                  {title()}
                  {description && (
                    <div className={classNames?.description}>
                      {description()}
                    </div>
                  )}
                </>
              )}
            </AccordionTrigger>
            <AccordionContent
              className={cn("mt-1 px-4 py-2", classNames?.content)}
            >
              {renderContent
                ? renderContent({ title, value, content })
                : content()}
            </AccordionContent>
          </AccordionItem>
        </AccordionRoot>
      ))}
    </div>
  );
};
