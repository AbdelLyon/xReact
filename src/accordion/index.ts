import "@/index.css";
export { Accordion, type AccordionItemData } from "@/accordion/Accordion";
