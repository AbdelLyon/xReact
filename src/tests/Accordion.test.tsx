import { render } from "@testing-library/react";
import { expect, it, describe } from "vitest";

import { Accordion, AccordionItemData } from "@/accordion";

describe("Accordion Component", () => {
  it("renders correctly with data and matches snapshot", () => {
    const items: AccordionItemData[] = [
      {
        title: () => <span>Item 1</span>,
        description: () => <span>Description 1</span>,
        value: "item-1",
        content: () => <div>Content 1</div>,
      },
      {
        title: () => <span>Item 2</span>,
        description: () => <span>Description 2</span>,
        value: "item-2",
        content: () => <div>Content 2</div>,
      },
    ];

    const { container } = render(<Accordion items={items} />);

    expect(container).toMatchSnapshot();
  });

  it("renders without data and matches snapshot", () => {
    const { container } = render(<Accordion items={[]} />);

    expect(container).toMatchSnapshot();
  });

  it("renders correctly with custom renderHeader", () => {
    const items: AccordionItemData[] = [
      {
        title: () => <span>Item 1</span>,
        description: () => <span>Description 1</span>,
        value: "item-1",
        content: () => <div>Content 1</div>,
      },
    ];

    const renderHeader = ({ title, description }: AccordionItemData) => (
      <div>
        <h3>{title()}</h3>
        {description && <p>{description()}</p>}
      </div>
    );

    const { container } = render(
      <Accordion items={items} renderHeader={renderHeader} />,
    );

    expect(container).toMatchSnapshot();
  });
});
