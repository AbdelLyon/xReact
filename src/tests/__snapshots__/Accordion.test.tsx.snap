// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Accordion Component > renders correctly with custom renderHeader 1`] = `
<div>
  <div
    class=""
  >
    <div
      class="w-full rounded-md bg-card-light dark:bg-card"
      data-orientation="vertical"
    >
      <div
        class="border border-border rounded-md"
        data-orientation="vertical"
        data-state="closed"
      >
        <h3
          class="relative flex"
          data-orientation="vertical"
          data-state="closed"
        >
          <button
            aria-controls="radix-:r5:"
            aria-expanded="false"
            class="flex flex-1 items-center justify-between py-4 font-medium transition-all [&[data-state=open]>svg]:rotate-180 pr-3 pl-1"
            data-orientation="vertical"
            data-radix-collection-item=""
            data-state="closed"
            id="radix-:r4:"
            type="button"
          >
            <div>
              <h3>
                <span>
                  Item 1
                </span>
              </h3>
              <p>
                <span>
                  Description 1
                </span>
              </p>
            </div>
            <svg
              class="lucide lucide-chevron-down -mt-5 size-4 shrink-0 transition-transform duration-200"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="m6 9 6 6 6-6"
              />
            </svg>
          </button>
        </h3>
        <div
          aria-labelledby="radix-:r4:"
          class="overflow-hidden text-sm transition-all data-[state=closed]:animate-accordion-up data-[state=open]:animate-accordion-down"
          data-orientation="vertical"
          data-state="closed"
          hidden=""
          id="radix-:r5:"
          role="region"
          style="--radix-accordion-content-height: var(--radix-collapsible-content-height); --radix-accordion-content-width: var(--radix-collapsible-content-width);"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`Accordion Component > renders correctly with data and matches snapshot 1`] = `
<div>
  <div
    class=""
  >
    <div
      class="w-full rounded-md bg-card-light dark:bg-card"
      data-orientation="vertical"
    >
      <div
        class="border border-border rounded-md"
        data-orientation="vertical"
        data-state="closed"
      >
        <h3
          class="relative flex"
          data-orientation="vertical"
          data-state="closed"
        >
          <button
            aria-controls="radix-:r1:"
            aria-expanded="false"
            class="flex flex-1 items-center justify-between py-4 font-medium transition-all [&[data-state=open]>svg]:rotate-180 pr-3 pl-1"
            data-orientation="vertical"
            data-radix-collection-item=""
            data-state="closed"
            id="radix-:r0:"
            type="button"
          >
            <span>
              Item 1
            </span>
            <div>
              <span>
                Description 1
              </span>
            </div>
            <svg
              class="lucide lucide-chevron-down -mt-5 size-4 shrink-0 transition-transform duration-200"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="m6 9 6 6 6-6"
              />
            </svg>
          </button>
        </h3>
        <div
          aria-labelledby="radix-:r0:"
          class="overflow-hidden text-sm transition-all data-[state=closed]:animate-accordion-up data-[state=open]:animate-accordion-down"
          data-orientation="vertical"
          data-state="closed"
          hidden=""
          id="radix-:r1:"
          role="region"
          style="--radix-accordion-content-height: var(--radix-collapsible-content-height); --radix-accordion-content-width: var(--radix-collapsible-content-width);"
        />
      </div>
    </div>
    <div
      class="w-full rounded-md bg-card-light dark:bg-card"
      data-orientation="vertical"
    >
      <div
        class="border border-border rounded-md"
        data-orientation="vertical"
        data-state="closed"
      >
        <h3
          class="relative flex"
          data-orientation="vertical"
          data-state="closed"
        >
          <button
            aria-controls="radix-:r3:"
            aria-expanded="false"
            class="flex flex-1 items-center justify-between py-4 font-medium transition-all [&[data-state=open]>svg]:rotate-180 pr-3 pl-1"
            data-orientation="vertical"
            data-radix-collection-item=""
            data-state="closed"
            id="radix-:r2:"
            type="button"
          >
            <span>
              Item 2
            </span>
            <div>
              <span>
                Description 2
              </span>
            </div>
            <svg
              class="lucide lucide-chevron-down -mt-5 size-4 shrink-0 transition-transform duration-200"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="m6 9 6 6 6-6"
              />
            </svg>
          </button>
        </h3>
        <div
          aria-labelledby="radix-:r2:"
          class="overflow-hidden text-sm transition-all data-[state=closed]:animate-accordion-up data-[state=open]:animate-accordion-down"
          data-orientation="vertical"
          data-state="closed"
          hidden=""
          id="radix-:r3:"
          role="region"
          style="--radix-accordion-content-height: var(--radix-collapsible-content-height); --radix-accordion-content-width: var(--radix-collapsible-content-width);"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`Accordion Component > renders without data and matches snapshot 1`] = `
<div>
  <div
    class=""
  />
</div>
`;
